### PR TITLE
chore: Checks PR to ensure new changelog entry file is present and valid

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Checks PR to ensure new changelog entry file is present and valid

Link to any related issue(s): CLOUDP-241728

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments

Example of validation error:
```
go run ./tools/check-changelog/*.go
2024/04/09 14:38:16 Error validating changelog file: .changelog/2126.txt, err: unknown changelog types [notes]: please use only the configured changelog entry types: [enhancement improvement feature bug note new-resource new-datasource new-function deprecation breaking-change]
exit status 1
make: *** [GNUmakefile:148: check-changelog] Error 1
```

### CODE REVIEW

**Comment by lantoli** [comment]
snyk error is a false positive, fix is already released

**Review by oarbusi — APPROVED** [review]
LGTM

**Review by corryroot — APPROVED** [review]
LGTM!

**Review by maastha — APPROVED** [review]

File: tools/check-changelog/check-changelog.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+	filePath := fmt.Sprintf(".changelog/%s.txt", number)
+	content, errFile := os.ReadFile(filePath)
+	if errFile == nil { // Always validate changelog file if present, skip logic is not considered in this case
+		entry := changelog.Entry{
+			Body: string(content),
+		}
+		if err := entry.Validate(); err != nil {
+			log.Fatalf("Error validating changelog file: %s, err: %v", filePath, err)
+		}
```
Comment (EspenAlbert) [inline tools/check-changelog/check-changelog.go]: [nit] could be a function `validChangelog(string(content))`

File: GNUmakefile [hunk 2]
```diff
@@ -138,7 +138,11 @@ update-tf-compatibility-matrix: ## Update Terraform Compatibility Matrix documen
 .PHONY: update-changelog-unreleased-section
 update-changelog-unreleased-section:
 	./scripts/update-changelog-unreleased-section.sh
-
+  
 .PHONY: generate-changelog-entry
 generate-changelog-entry:
 	./scripts/generate-changelog-entry.sh
+
+.PHONY: check-changelog
```
Comment (AgustinBettati) [inline GNUmakefile]: nit: would `check-changelog-entry-files` be more explicit?

**Review by EspenAlbert — APPROVED** [review]
Nice 👏

File: contributing/changelog-process.md [hunk 3]
```diff
@@ -143,3 +144,10 @@ make generate-changelog-entry
 ```
 
 - The `subcategory` input prompt refers to the prefix of the changelog entry, used for specifying the relevant resource/data source when needed (e.g. data-source/mongodbatlas_project)
+
+## PR Changelog check
+
+A PR check is included to validate the changelog file.
```
Comment (AgustinBettati) [inline contributing/changelog-process.md]:
```suggestion
A PR check is included to validate the changelog entry file. 
```

File: contributing/changelog-process.md [hunk 4]
```diff
@@ -143,3 +144,10 @@ make generate-changelog-entry
 ```
 
 - The `subcategory` input prompt refers to the prefix of the changelog entry, used for specifying the relevant resource/data source when needed (e.g. data-source/mongodbatlas_project)
+
+## PR Changelog check
+
+A PR check is included to validate the changelog file. 
+If a PR doesn't need a changelog its check can be skipped:
```
Comment (AgustinBettati) [inline contributing/changelog-process.md]:
```suggestion
If a PR doesn't need a changelog entry its check can be skipped:
```

File: tools/check-changelog/check-changelog.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		fmt.Println("Skipping changelog check because PR title")
+		return
+	}
+
+	if skipLabel(labels) {
+		fmt.Printf("Skipping changelog check because PR label found: %s\n", skipLabelName)
+		return
+	}
+
+	log.Fatalf("Consider using label %s if this PR doesn't need a changelog.\nChangelog file not found: %s, err: %v", skipLabelName, filePath, errFile)
```
Comment (AgustinBettati) [inline tools/check-changelog/check-changelog.go]: Would it be worth logging an additional line referencing to the guidelines? https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/changelog-process.md#changes-that-should-have-a-changelog-entry

**Review by AgustinBettati — APPROVED** [review]
LGTM, just minor suggestions

File: tools/check-changelog/check-changelog.go (hunk 1 above)
Comment (lantoli) [reply tools/check-changelog/check-changelog.go:40]: let me think more about that, the thing is that there are many variables in and out

File: GNUmakefile (hunk 2 above)
Comment (lantoli) [reply GNUmakefile]: sounds good, i'll change to check-changelog-entry-file in singular

Comment (lantoli) [reply GNUmakefile]: done

File: tools/check-changelog/check-changelog.go (hunk 5 above)
Comment (lantoli) [reply tools/check-changelog/check-changelog.go]: if it's ok I'll add a sentence to suggest reviewing the contribution guide as the link is very long and might change and get outdated

File: tools/check-changelog/check-changelog.go (hunk 1 above)
Comment (lantoli) [reply tools/check-changelog/check-changelog.go]: done, thanks

**Pull request merged**